File: agent/agent.py
"""
AgentWise AI Agent — Autonomous Savings Monitor
================================================
A lightweight Python agent that:
  1. Polls a mock income source (or real XRPL account) for incoming payments.
  2. When income >= threshold, calls the AgentWise API to trigger an auto top-up
     (opens a state channel if needed, or executes a deposit).
  3. Logs all actions for auditability.

Usage:
    cp ../.env.example .env  # fill in AGENT_* vars
    python agent.py

For production, replace mock_check_income() with real XRPL account monitoring.
"""

import os
import time
import logging
import httpx
from dotenv import load_dotenv

load_dotenv()

# ── Config ─────────────────────────────────────────────────────────────────────

API_URL: str = os.getenv("NEXT_PUBLIC_API_URL", "http://localhost:4000")
GOAL_ID: str = os.getenv("AGENT_GOAL_ID", "")
PARTICIPANT_ADDRESS: str = os.getenv("AGENT_XRPL_ADDRESS", "")
TOP_UP_THRESHOLD: float = float(os.getenv("AGENT_TOP_UP_THRESHOLD_RLUSD", "10"))
POLL_INTERVAL: int = int(os.getenv("AGENT_POLL_INTERVAL_SECONDS", "30"))
XRPL_NODE: str = os.getenv("NEXT_PUBLIC_XRPL_NODE_TESTNET", "wss://s.altnet.rippletest.net:51233")

# ── Logger ─────────────────────────────────────────────────────────────────────

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [AgentWise Agent] %(levelname)s  %(message)s",
    datefmt="%Y-%m-%dT%H:%M:%S",
)
log = logging.getLogger("agentwise")


# ── Mock Income Source ─────────────────────────────────────────────────────────

def mock_check_income() -> float:
    """
    Simulate checking for new income.

    In production, replace with:
      - xrpl-py account subscription to detect incoming payments
      - Integration with payroll / freelance platform webhooks
      - Chipper Cash or Yellow Card balance polling

    Returns the detected income amount in RLUSD (0.0 if none).
    """
    import random
    # Simulate income arriving ~40% of the time, between ₦500–₦5000 (mock RLUSD)
    if random.random() < 0.4:
        amount = round(random.uniform(5.0, 50.0), 2)
        log.info(f"Income detected: {amount} RLUSD")
        return amount
    return 0.0


# ── API Calls ──────────────────────────────────────────────────────────────────

def get_goal_status(client: httpx.Client) -> dict:
    """Fetch current goal status from the AgentWise API."""
    resp = client.get(f"{API_URL}/api/agent/status/{GOAL_ID}", timeout=10)
    resp.raise_for_status()
    return resp.json()


def trigger_top_up(client: httpx.Client, income_amount: float) -> dict:
    """
    Signal the AgentWise API to execute a deposit or open a channel.
    The API handles the logic of whether to deposit off-chain or open a new channel.
    """
    payload = {
        "goalId": GOAL_ID,
        "incomeAmount": income_amount,
        "participantAddress": PARTICIPANT_ADDRESS,
        "additionalFunding": round(income_amount * 0.1, 4),  # auto-save 10% of income
    }
    resp = client.post(f"{API_URL}/api/agent/top-up", json=payload, timeout=15)
    resp.raise_for_status()
    return resp.json()


# ── Main Loop ──────────────────────────────────────────────────────────────────

def run_agent() -> None:
    """Main polling loop."""
    log.info("=" * 60)
    log.info("AgentWise Autonomous Savings Agent starting…")
    log.info(f"  API:              {API_URL}")
    log.info(f"  Goal ID:          {GOAL_ID or '⚠ NOT SET'}")
    log.info(f"  Participant:      {PARTICIPANT_ADDRESS or '⚠ NOT SET'}")
    log.info(f"  Top-up threshold: {TOP_UP_THRESHOLD} RLUSD")
    log.info(f"  Poll interval:    {POLL_INTERVAL}s")
    log.info("=" * 60)

    if not GOAL_ID:
        log.error("AGENT_GOAL_ID not set. Create a goal first, then set it in .env")
        return

    consecutive_errors = 0

    with httpx.Client() as client:
        while True:
            try:
                # 1. Check current goal status
                status = get_goal_status(client)
                progress = status.get("progress", {})
                log.info(
                    f"Goal: {status['goal']['name']} | "
                    f"Saved: {progress.get('savedAmount', 0):.2f} / "
                    f"{progress.get('targetAmount', 0):.2f} RLUSD "
                    f"({progress.get('percentComplete', 0):.1f}%)"
                )

                # 2. Check for income
                income = mock_check_income()

                # 3. Trigger top-up if above threshold
                if income >= TOP_UP_THRESHOLD:
                    log.info(f"Income {income} RLUSD >= threshold {TOP_UP_THRESHOLD}. Triggering top-up…")
                    result = trigger_top_up(client, income)
                    action = result.get("action", "unknown")
                    log.info(f"Top-up result: {action}")

                    if action == "deposit_executed":
                        deposit = result.get("deposit", {})
                        log.info(
                            f"  Deposit: {deposit.get('amount', 0):.2f} RLUSD "
                            f"(channel: {deposit.get('channelId', 'n/a')})"
                        )
                    elif action == "channel_opened":
                        log.info(f"  New channel: {result.get('channelId')}")

                    after_pct = result.get("progress", {}).get("percentComplete", 0)
                    log.info(f"  Progress after: {after_pct:.1f}%")
                else:
                    log.debug(f"Income {income:.2f} RLUSD below threshold. No action.")

                consecutive_errors = 0

            except httpx.HTTPStatusError as e:
                log.error(f"API error {e.response.status_code}: {e.response.text[:200]}")
                consecutive_errors += 1
            except httpx.RequestError as e:
                log.error(f"Network error: {e}")
                consecutive_errors += 1
            except Exception as e:
                log.error(f"Unexpected error: {e}")
                consecutive_errors += 1

            # Back-off on repeated errors
            if consecutive_errors >= 5:
                backoff = POLL_INTERVAL * 4
                log.warning(f"5 consecutive errors. Backing off {backoff}s…")
                time.sleep(backoff)
                consecutive_errors = 0
            else:
                time.sleep(POLL_INTERVAL)


# ── Entry Point ────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    run_agent()

File: agent/agent.mjs
/**
 * AgentWise AI Agent — Node.js version
 * =====================================
 * Equivalent to agent.py but runs on Node.js (no Python needed).
 * 
 * Usage:
 *   node agent.mjs
 *
 * Set environment variables (or create agent/.env):
 *   NEXT_PUBLIC_API_URL      - API base URL (default: http://localhost:4000)
 *   AGENT_GOAL_ID            - Goal ID to monitor
 *   AGENT_XRPL_ADDRESS       - Participant XRPL address
 *   AGENT_TOP_UP_THRESHOLD_RLUSD - Min income to trigger top-up (default: 10)
 *   AGENT_POLL_INTERVAL_MS   - Poll interval in ms (default: 5000)
 */

import { readFileSync, existsSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

// ── Load .env manually (no dotenv dep needed in Node 20+) ────────────────────
const __dir = dirname(fileURLToPath(import.meta.url));
const envFile = join(__dir, ".env");
if (existsSync(envFile)) {
  const lines = readFileSync(envFile, "utf8").split("\n");
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const [key, ...rest] = trimmed.split("=");
    if (key && rest.length) process.env[key.trim()] = rest.join("=").trim();
  }
}

// ── Config ────────────────────────────────────────────────────────────────────
const API_URL          = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:4000";
const GOAL_ID          = process.env.AGENT_GOAL_ID ?? "";
const PARTICIPANT_ADDR = process.env.AGENT_XRPL_ADDRESS ?? "";
const THRESHOLD        = parseFloat(process.env.AGENT_TOP_UP_THRESHOLD_RLUSD ?? "10");
const POLL_MS          = parseInt(process.env.AGENT_POLL_INTERVAL_MS ?? "5000", 10);

// ── Logger ────────────────────────────────────────────────────────────────────
function log(level, msg) {
  const ts = new Date().toISOString().replace("T", " ").slice(0, 19);
  console.log(`${ts} [AgentWise] ${level.padEnd(5)}  ${msg}`);
}

// ── Mock income detector ──────────────────────────────────────────────────────
function checkIncome() {
  // Simulate income arriving ~40% of the time, 5–50 RLUSD
  if (Math.random() < 0.4) {
    const amount = parseFloat((5 + Math.random() * 45).toFixed(2));
    log("INFO", `Income detected: ${amount} RLUSD`);
    return amount;
  }
  return 0;
}

// ── API calls ─────────────────────────────────────────────────────────────────
async function getStatus() {
  const res = await fetch(`${API_URL}/api/agent/status/${GOAL_ID}`);
  if (!res.ok) throw new Error(`Status fetch failed: ${res.status}`);
  return res.json();
}

async function triggerTopUp(incomeAmount) {
  const res = await fetch(`${API_URL}/api/agent/top-up`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      goalId: GOAL_ID,
      incomeAmount,
      participantAddress: PARTICIPANT_ADDR,
      additionalFunding: parseFloat((incomeAmount * 0.1).toFixed(4)),
    }),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Top-up failed ${res.status}: ${body}`);
  }
  return res.json();
}

// ── Main loop ─────────────────────────────────────────────────────────────────
async function run() {
  console.log("=".repeat(60));
  log("INFO", "AgentWise Autonomous Savings Agent (Node.js) starting...");
  log("INFO", `API:              ${API_URL}`);
  log("INFO", `Goal ID:          ${GOAL_ID || "⚠ NOT SET"}`);
  log("INFO", `Participant:      ${PARTICIPANT_ADDR || "⚠ NOT SET"}`);
  log("INFO", `Threshold:        ${THRESHOLD} RLUSD`);
  log("INFO", `Poll interval:    ${POLL_MS}ms`);
  console.log("=".repeat(60));

  if (!GOAL_ID) {
    log("ERROR", "AGENT_GOAL_ID not set. Create a goal first via the dashboard.");
    process.exit(1);
  }

  let errors = 0;
  let cycles = 0;

  while (true) {
    cycles++;
    try {
      // 1. Get current goal status
      const status = await getStatus();
      const p = status.progress;
      log("INFO", `[Cycle ${cycles}] ${status.goal.name} | Saved: ${p.savedAmount.toFixed(2)}/${p.targetAmount.toFixed(2)} RLUSD (${p.percentComplete.toFixed(1)}%) | ${p.daysRemaining}d left`);

      // 2. Check for income
      const income = checkIncome();

      // 3. Act if above threshold
      if (income >= THRESHOLD) {
        log("INFO", `Income ${income} >= threshold ${THRESHOLD}. Triggering top-up...`);
        const result = await triggerTopUp(income);
        const action = result.action;

        if (action === "deposit_executed") {
          log("INFO", `  ✓ Deposit: ${result.deposit.amount} RLUSD | channel: ${result.deposit.channelId}`);
          log("INFO", `  → Progress: ${result.progress.percentComplete.toFixed(1)}%`);
        } else if (action === "channel_opened") {
          log("INFO", `  ✓ New channel opened: ${result.channelId}`);
        } else {
          log("INFO", `  Action: ${action}`);
        }

        if (p.percentComplete >= 100) {
          log("INFO", "🎉 Goal COMPLETED! Agent shutting down.");
          break;
        }
      } else {
        log("DEBUG", `Income ${income.toFixed(2)} below threshold. No action.`);
      }

      errors = 0;
    } catch (err) {
      errors++;
      log("ERROR", `${err.message} (consecutive errors: ${errors})`);
      if (errors >= 5) {
        const backoff = POLL_MS * 4;
        log("WARN", `5 consecutive errors. Backing off ${backoff}ms...`);
        await new Promise(r => setTimeout(r, backoff));
        errors = 0;
        continue;
      }
    }

    await new Promise(r => setTimeout(r, POLL_MS));
  }
}

run().catch(err => {
  console.error("Fatal:", err);
  process.exit(1);
});
